### PR TITLE
 #515 Improve Retry tests

### PR DESCRIPTION
**Pull request by sebady**
Change the java RetryTest to be a junit test
Also minor test additions to RetrySpec

Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [ ] Documentation added/updated.
- [ ] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: squbs-ext/src/test/java/org/squbs/streams/RetryTest.java [hunk 1]
```diff
@@ -26,8 +26,8 @@
 import akka.stream.javadsl.Sink;
 import akka.stream.javadsl.Source;
 import akka.stream.testkit.javadsl.TestSink;
+import org.junit.Test;
```
Comment (akara) [inline squbs-ext/src/test/java/org/squbs/streams/RetryTest.java]: Please make sure the assertions are also JUnit assertions. The signature of these two between expect and actual is inverted. So your assert calls need to be changed.

**Comment by akara** [comment]
Can you please let me know the test coverage you got after this change? Also, is this PR ready to be merged?

**Comment by sebady** [comment]
The scoverage report shows 85% line coverage

**Comment by anilgursel** [comment]
Comments for [this coverage report](https://github.com/paypal/squbs/files/1953019/Retry.scala.html.log) (remove the .log extension and open in a browser, apparently Github does not support html):

Line 140: There should be a corresponding test case in CircuitBreaker, you can take a look at there.

Line 170: Is there any scenario where it can really be `None`?  Given that we are anyway throwing an exception, we can instead just do a `get` on the `Option`.

Line 217: We can reproduce this scenario by causing retries with really long delays, while elements are still in registry, a stage on the upstream can fail.  However, I am not really sure about the logic here.  In any scenario we fail. Shouldn't we let let the messages be drained?

Line 229: This one might be tricky to reproduce consistently.  Thinking.. 🤔   I wonder if adding a buffer stage on the upstream, or making the internal buffer size high(more than the internal registry size) may help reproduce.

Line 240: After making sure there is no element in the registry, completing a stage in the joined flow should trigger this.  However, it needs to be tested that whether `onDowstreamFinish` of `out1` or `onUpstreamFinish` of `in2` is called, or indeterministic.

Line 255: Any scenario where we keep retrying and reaching to the registry size limit and ultimately causing back pressure, and resuming after a while should cover this line.

Line 263: This will never get called.  I suggest removing this in the next or next-to-next release.  interesting thing though, how come the comment line shows green 😄 

Line 269: When exponential back off is in place, and let's say 2nd element is failing second time with a really long delay, and 3rd (or later) element comes in and will be re-tried very soon..  such scenarios should cover this.

Line 275: Let's say, only a single element is going through the stream and it is no-delay and a failure happens.  It should cover this branch.  We should also be able to reproduce this with multiple elements and a large buffer in the joined flow (the buffer would make sure out1 is always available).

Line 293: Similar to 255.  

Line 300: Same as 263.

Line 310: While all elements are in registry, if an upstream stage completes, that should ultimately trigger this scenario.

Line 464: Don't we already have JMX tests?  I would assume those would cover it.

**Comment by sebady** [comment]
Line 170: The only scenario where InputBuffer would not exist in the `inheritedAttributes` is probably user error.  So I can change this to a simpler `inheritedAttributes.mandatoryAttribute[InputBuffer].max)` 

Line 217: I think  the current implementation of `onUpstreamFailure` that either "propagates" the exception on `out1` or fails the stage seems like reasonable behavior for this stage.
Line 255 and 269:  Still trying to add test for this scenario.

Line 464: Should now be covered in latest commit on PR.

**Pull request merged**